hal/ia32/memory.c

```diff
@@ -195,28 +195,64 @@ int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
 
 
 /* Flushes 8042 keyboard controller buffers */
-static void memory_empty8042(void)
+static int memory_empty8042(void)
 {
 	unsigned char status;
-
-	do {
+	unsigned short int timeout;
+	for (timeout = 0xffff; timeout > 0; --timeout) {
 		/* Discard input data */
-		if ((status = hal_inb((void *)0x64)) & 0x01) {
+		status = hal_inb((void *)0x64);
+		if ((status & 0x01) != 0) {
 			hal_inb((void *)0x60);
 			continue;
 		}
-	} while (status & 0x02);
+		if ((status & 0x02) == 0) {
+			return 0;
+		}
+	}
+	return -1;
+}
+
+
+static int checkA20Line(void)
+{
+	int result = 0;
+	/* Set base to some address in the first 1MiB that can be overwritten */
+	volatile u16 *const base = (void *)0x7dfe; /* Address of the magic number at the end of MBR (0xAA55) */
```

Comment:
`ADDR_PLO` is `0x7c00` where MBR (one sector 512 bytes) is loaded.
```suggestion
	/* Address of the magic number at the end of MBR (0xAA55) */
	volatile u16 *const base = (void *)((ADDR_PLO) + 512 - 2);
```